src/GraphQL/Internal/Validation.hs

```diff
@@ -1,104 +1,416 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+-- | Transform GraphQL query documents from AST into valid structures
+--
+-- This corresponds roughly to the
+-- [Validation](https://facebook.github.io/graphql/#sec-Validation) section of
+-- the specification, except where noted.
+--
+-- One core difference is that this module doesn't attempt to do any
+-- type-level validation, as we attempt to defer all of that to the Haskell
+-- type checker.
+--
+-- Still missing:
+--
+--   * variable validation <https://facebook.github.io/graphql/#sec-Validation.Variables>
+--   * field selection merging <https://facebook.github.io/graphql/#sec-Field-Selection-Merging>
+--   * input object field uniqueness <https://facebook.github.io/graphql/#sec-Values>
+--
+-- Deliberately not going to do:
+--
+--   * field selections on compound types <https://facebook.github.io/graphql/#sec-Field-Selections-on-Objects-Interfaces-and-Unions-Types>
+--   * leaf field selections <https://facebook.github.io/graphql/#sec-Leaf-Field-Selections>
+--   * argument names <https://facebook.github.io/graphql/#sec-Argument-Names>
+--   * argument value type correctness <https://facebook.github.io/graphql/#sec-Argument-Values-Type-Correctness>
+--   * fragment spread type existence <https://facebook.github.io/graphql/#sec-Fragment-Spread-Type-Existence>
+--   * fragments on compound types <https://facebook.github.io/graphql/#sec-Fragments-On-Composite-Types>
+--   * fragment spread is possible <https://facebook.github.io/graphql/#sec-Fragment-spread-is-possible>
+--   * directives are defined <https://facebook.github.io/graphql/#sec-Directives-Are-Defined>
+--   * directives are in valid locations <https://facebook.github.io/graphql/#sec-Directives-Are-In-Valid-Locations>
+--   * variable default values are correctly typed <https://facebook.github.io/graphql/#sec-Variable-Default-Values-Are-Correctly-Typed>
+--   * variables are input types <https://facebook.github.io/graphql/#sec-Variables-Are-Input-Types>
+--   * all variable usages are allowed <https://facebook.github.io/graphql/#sec-All-Variable-Usages-are-Allowed>
+--
+-- Because all of the above rely on type checking.
 module GraphQL.Internal.Validation
   ( ValidationError(..)
-  , ValidDocument
+  , QueryDocument
   , validate
   , getErrors
+  -- * Operating on validated documents
+  , getOperation
   -- * Exported for testing
   , findDuplicates
   ) where
 
 import Protolude
 
+import Data.List.NonEmpty (NonEmpty(..))
+import qualified Data.List.NonEmpty as NonEmpty
+import qualified Data.Map as Map
+import qualified Data.Set as Set
 import qualified GraphQL.Internal.AST as AST
+-- Directly import things from the AST that do not need validation, so that
+-- @AST.Foo@ in a type signature implies that something hasn't been validated.
+import GraphQL.Internal.AST (Name, Alias, TypeCondition)
+
+-- | A valid query document.
+--
+-- Construct this using 'validate' on an 'AST.QueryDocument'.
+data QueryDocument
```

Comment:
The difference is "need" vs "want". `Server.hs` operates on AST, and AST is the same as it has been since #45, so we can change this with impunity.

However, we really should be using this stuff, and we can after this PR is done.